src/middleware/routeHandler.ts

```diff
@@ -0,0 +1,262 @@
+    originalStagingLiteCommitSha = stagingLiteResult.value.sha
+    if (!originalStagingCommitSha || !originalStagingLiteCommitSha) {
+      await unlock(siteName)
+      if (stagingResult.isErr()) next(stagingResult.error)
+      return
+    }
+    // Unused for git file system, but to maintain existing structure
+    res.locals.githubSessionData = new GithubSessionData({
+      currentCommitSha: "",
+      treeSha: "",
+    })
+  } else {
+    try {
+      const {
+        currentCommitSha: currentStgCommitSha,
+        treeSha: stgTreeSha,
+      } = await getCommitAndTreeSha(siteName, accessToken, STAGING_BRANCH)
+
+      const {
+        currentCommitSha: currentStgLiteCommitSha,
+      } = await getCommitAndTreeSha(siteName, accessToken, STAGING_LITE_BRANCH)
+
+      const githubSessionData = new GithubSessionData({
+        currentCommitSha: currentStgCommitSha,
+        treeSha: stgTreeSha,
+      })
+      res.locals.githubSessionData = githubSessionData
+
+      originalStagingCommitSha = currentStgCommitSha
+      originalStagingLiteCommitSha = currentStgLiteCommitSha
+    } catch (err) {
+      await unlock(siteName)
+      next(err)
+      return
+    }
+  }
+
+  await routeHandler(req, res, next).catch(async (err: any) => {
+    try {
+      if (shouldUseGitFileSystem) {
+        await backOff(
+          () =>
+            convertNeverThrowToPromise(
+              gitFileSystemService.rollback(
+                siteName,
+                originalStagingCommitSha,
+                STAGING_BRANCH
+              )
+            ),
+          backoffOptions
+        )
+
+        await backOff(
+          () =>
+            convertNeverThrowToPromise(
+              gitFileSystemService.rollback(
+                siteName,
+                originalStagingLiteCommitSha,
+                STAGING_LITE_BRANCH
+              )
+            ),
+          backoffOptions
+        )
+
+        await backOff(() => {
+          let pushRes = gitFileSystemService.push(
+            siteName,
+            STAGING_BRANCH,
+            true
+          )
+          if (originalStagingLiteCommitSha) {
+            pushRes = pushRes.andThen(() =>
+              gitFileSystemService.push(siteName, STAGING_LITE_BRANCH, true)
+            )
+          }
+
+          return convertNeverThrowToPromise(pushRes)
+        }, backoffOptions)
+      } else {
+        await backOff(
+          () =>
+            revertCommit(
+              originalStagingCommitSha,
+              siteName,
+              accessToken,
+              STAGING_BRANCH
+            ),
+          backoffOptions
+        )
+        await backOff(
+          () =>
+            revertCommit(
+              originalStagingLiteCommitSha,
+              siteName,
+              accessToken,
+              STAGING_LITE_BRANCH
+            ),
+          backoffOptions
+        )
+      }
+    } catch (retryErr) {
+      await unlock(siteName)
```

Comment:
tbvh this whole chunk is very jank and confusing but this PR doesn't focus on code clarity but migration from js -> ts so ok with it. we should, however, look into probably refactoring this into a sensible form.